File: apps/cli/Commands/Pull/SecretsPuller.cs
using DepVault.Cli.Auth;
using DepVault.Cli.Crypto;
using DepVault.Cli.Output;
using Spectre.Console;
using VaultModel = DepVault.Cli.ApiClient.Api.Projects.Item.Vaults.Vaults;

namespace DepVault.Cli.Commands.Pull;

/// <summary>Lists and downloads secret files for selected vaults.</summary>
public sealed class SecretsPuller(
    IApiClientFactory clientFactory,
    IOutputFormatter output)
{
    /// <summary>Downloads secret files for the selected vaults. Returns number of files written.</summary>
    public async Task<int> PullAsync(
        string projectId, List<VaultModel> vaults,
        string outputDir, byte[] dek, CancellationToken ct)
    {
        var client = clientFactory.Create();
        var selectedVaultIds = vaults.Select(v => v.Id).ToHashSet();

        ApiClient.Api.Projects.Item.Secrets.SecretsGetResponse? files;

        try
        {
            files = await AnsiConsole.Status()
                .Spinner(Spinner.Known.Dots)
                .StartAsync("Fetching secret files...", async _ =>
                    await client.Api.Projects[projectId].Secrets.GetAsync(config =>
                    {
                        config.QueryParameters.Page = 1;
                        config.QueryParameters.Limit = 100;
                    }, ct));
        }
        catch (Exception ex)
        {
            output.PrintError($"Failed to list secret files: {ex.Message}");
            return 0;
        }

        var items = files?.Items?
            .Where(f => selectedVaultIds.Contains(f.VaultId))
            .ToList();

        if (items is null || items.Count == 0)
        {
            AnsiConsole.MarkupLine("[grey]No secret files found.[/]");
            return 0;
        }

        var filesWritten = 0;

        foreach (var file in items)
        {
            try
            {
                var vault = vaults.FirstOrDefault(v => v.Id == file.VaultId);
                var secretsDir = ResolveSecretsDir(vault, outputDir);
                Directory.CreateDirectory(secretsDir);

                var filePath = Path.Combine(secretsDir, file.Name ?? $"secret-{file.Id}");

                var downloadResult = await client.Api.Projects[projectId].Secrets[file.Id!].Download
                    .GetAsync(cancellationToken: ct);

                if (downloadResult is null || string.IsNullOrEmpty(downloadResult.EncryptedContent))
                {
                    output.PrintError($"Failed to download {file.Name}: empty response");
                    continue;
                }

                var decryptedBytes = VaultCrypto.DecryptBytes(
                    downloadResult.EncryptedContent, downloadResult.Iv ?? "",
                    downloadResult.AuthTag ?? "", dek);
                await File.WriteAllBytesAsync(filePath, decryptedBytes, ct);

                output.PrintSuccess($"  {Path.GetRelativePath(outputDir, filePath)}");
                filesWritten++;
            }
            catch (Exception ex)
            {
                output.PrintError($"Failed to download {file.Name}: {ex.Message}");
            }
        }

        return filesWritten;
    }

    private static string ResolveSecretsDir(VaultModel? vault, string outputDir)
    {
        var dirPath = vault?.DirectoryPath;

        if (!string.IsNullOrEmpty(dirPath))
        {
            return Path.Combine(outputDir, dirPath, ".secrets");
        }

        return Path.Combine(outputDir, ".secrets");
    }
}

File: apps/cli-ts/src/services/secrets-puller.ts
import { decryptBinary, fromBase64 } from "@depvault/crypto";
import { getApiClient } from "./api-client";

interface SecretFileResult {
  name: string;
  content: ArrayBuffer;
}

/** Downloads and decrypts a secret file. */
export async function pullSecretFile(
  projectId: string,
  fileId: string,
  dek: CryptoKey,
): Promise<SecretFileResult> {
  const client = getApiClient();
  const { data, error } = await client.api
    .projects({ id: projectId })
    .secrets({ fileId })
    .download.get();

  if (error || !data) {
    throw new Error(`Failed to download secret file ${fileId}.`);
  }

  const file = data;
  const content = await decryptBinary(file.encryptedContent, file.iv, file.authTag, dek);

  return { name: file.name, content };
}

/** Lists all secret files for a project (metadata only). */
export async function listSecretFiles(projectId: string) {
  const client = getApiClient();
  const { data, error } = await client.api
    .projects({ id: projectId })
    .secrets.get({ query: { page: 1, limit: 100 } });

  if (error || !data) {
    throw new Error("Failed to list secret files.");
  }

  return (data.items ?? []).map((s) => ({
    id: s.id,
    name: s.name,
    vaultGroupName: s.vaultGroupName,
    fileSize: s.fileSize,
    mimeType: s.mimeType,
    createdAt: s.createdAt,
  }));
}
